scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/LiveJobs.java

```diff
@@ -423,18 +463,59 @@ TerminationData taskTerminatedWithResult(TaskId taskId, TaskResultImpl result) {
+
+                int numberOfExecutionLeft = task.getNumberOfExecutionLeft();
+
+                if (numberOfExecutionLeft <= 0 && onErrorPolicyInterpreter.requiresCancelJobOnError(task)) {
+                    logger.info("No retry left and task is tagged with cancel job on error");
+
+                    jobData.job.increaseNumberOfFaultyTasks(taskId);
                     endJob(jobData, terminationData, task, result,
-                            "An error occurred in your task and the maximum number of executions has been reached. "
-                                + "You also ask to cancel the job in such a situation !", JobStatus.CANCELED);
-                    return terminationData;
-                } else if (task.getNumberOfExecutionLeft() > 0) {
-                    long waitTime = jobData.job.getNextWaitingTime(task.getMaxNumberOfExecution() -
-                        task.getNumberOfExecutionLeft());
-                    restartTaskOnError(jobData, task, TaskStatus.WAITING_ON_ERROR, result, waitTime,
-                            terminationData);
+                            "An error occurred in your task and the maximum number of executions has been reached. " +
+                                    "You also ask to cancel the job in such a situation!",
+                            JobStatus.CANCELED);
+
+                    logger.info("Job has been canceled");
+
                     return terminationData;
+                } else if (numberOfExecutionLeft > 0) {
+                    logger.info("Number of execution left is " + numberOfExecutionLeft);
+
+                    if (onErrorPolicyInterpreter.requiresPauseTaskOnError(task)) {
+                        suspendTaskOnError(jobData, task, result.getTaskDuration());
+
+                        logger.info("Task is paused on error");
+
+                        return terminationData;
+                    } else if (requiresPauseJobOnError) {
+                        suspendTaskOnError(jobData, task, result.getTaskDuration());
+                        pauseJob(task.getJobId());
+
+                        logger.info("Job is paused on error");
+
+                        return terminationData;
+                    } else {
+                        jobData.job.increaseNumberOfFaultyTasks(taskId);
+
+                        long waitTime = jobData.job
+                                .getNextWaitingTime(task.getMaxNumberOfExecution() - numberOfExecutionLeft);
+                        restartTaskOnError(jobData, task, TaskStatus.WAITING_ON_ERROR, result, waitTime,
+                                terminationData);
+
+                        logger.info("New restart is scheduled");
+
+                        return terminationData;
+                    }
+                } else if (numberOfExecutionLeft <= 0) {
+                    jobData.job.increaseNumberOfFaultyTasks(taskId);
+
+                    if (requiresPauseJobOnError) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS134)